Create and test Node and ItemNode

sydneyli opened the pull request:
Fleshing out API as described in #49. Starting with generic Node and a more specific ItemNode implementation.

etingof commented on apacheconfig/wloader.py:
```diff
@@ -0,0 +1,160 @@
+        but has a couple different possible `type`s, including `comment`,
+        `statement`, and `include`-- depending on which the caller may
+        decide to treat ItemNode differently.
+        """
```
/me notes that we should properly document parameters, return value, exceptions etc...

etingof commented on apacheconfig/wloader.py:
```diff
@@ -0,0 +1,160 @@
+    def type(self):
+        """A typestring as defined by the apacheconfig parser.
+
+        A single *Node class can have multiple possible values for this. For
```
Should we have these values enumerated somewhere?

sydneyli replied:
I just did this in the documentation for the time being, but in the future maybe we can do this with #68 as well.

etingof commented on apacheconfig/wloader.py:
```diff
@@ -0,0 +1,160 @@
+
+    @property
+    def whitespace(self):
+        """Trailing or preceding whitespace.
```
It is not clear from the docstring how many whitespaces will be reported. If it's just one, how the rest would be represented?

etingof commented on apacheconfig/wloader.py:
```diff
@@ -0,0 +1,160 @@
+
+
+class ItemNode(Node):
+    """Contains data for a single configuration line.
```
This reads confusingly to me. This Apache syntax does not seem to be heavily line-oriented. You can have multi-line comments, tags etc. So what exactly `ItemNode` represents?

etingof commented on apacheconfig/wloader.py:
```diff
@@ -0,0 +1,160 @@
+        ['comment', '# here is a comment']
+        ['comment', '\n  ', '# here is a comment']
+
+    Properties:
```
nit: perhaps we could documents them as their own docstrings. Sphinx would pull them from there just fine.

etingof commented on apacheconfig/wloader.py:
```diff
@@ -0,0 +1,160 @@
+    """
+
+    def __init__(self, raw):
+        """Initializes ItemNode with raw data from AST
```
Arguably, putting `__init__` docstring into class docstring makes it looking better in Sphinx.

etingof commented on apacheconfig/wloader.py:
```diff
@@ -0,0 +1,160 @@
+                "".join([_restore_original(word) for word in self._raw]))
+
+    def __repr__(self):
+        return ("ItemNode(%s)"
```
nit: `self.__class__.__name__` or something alike would avoid hardcoding class name.

sydneyli commented on apacheconfig/wloader.py:
```diff
@@ -0,0 +1,160 @@
+
+    @abc.abstractmethod
+    def __str__(self):
+        raise NotImplementedError()
```
Yes, you're right-- I think a `pass` should suffice.

etingof replied:
yeah, `pass` or a doctstring where is makes sense

sydneyli commented on apacheconfig/wloader.py:
```diff
@@ -0,0 +1,160 @@
+        raise NotImplementedError()
+
+    @property
+    def type(self):
```
I think it should be fine since it's tied to the object-- but to be extra safe I'll rename it `parser_type`!

sydneyli commented:
Thanks for the feedback! I addressed some of the comments-- I'll try to address the rest of them by Sphinx-ifying the docs.

sydneyli commented:
Documentation/testing sprint done! While doing that I also changed the constructor a bit-- I moved the `parse_item` factory as a static method at `ItemNode.parse`.

I pushed a branch at `sydneyli/apacheconfig:merged-writable-loader-item-sphinx` that merges this branch with #77 to show that the docs make the `apacheconfig.wloader` automodule properly.

etingof commented on apacheconfig/wloader.py:
```diff
@@ -30,98 +28,149 @@ class Node():
+        """Writes this node to a raw string. To get more metadata about the
+        object, use ``repr``.
+        """
+        pass
```
nit: in presence of a docstring, `pass` can be omitted

nit++: both `__str__` and `__repr__` have predefined semantics: the first prints the object prettily, the second (ideally) produces valid Python code which can be evaluated and re-produce a copy of the original object. 

That gives me the impression that:

1. it sometimes makes sense to stick with this pre-defined meanings whenever possible
2. if (1) holds, the docstring is not really necessary because everyone expects what these methods would do

Tell me when you can't tolerate my nitpicking any longer!

sydneyli replied:
These are great comments! I'm not fully familiar with Python best practices, so they're very helpful. What I'll do here then is:

1. Move current functionality of `str` to a new function called `dump()`
2. Move functionality of `__repr__` to `__str__`
3. Remove docstsring

etingof commented on apacheconfig/wloader.py:
```diff
@@ -30,98 +28,149 @@ class Node():
-        but has a couple different possible `type`s, including `comment`,
-        `statement`, and `include`-- depending on which the caller may
-        decide to treat ItemNode differently.
+        The possible values for this are:
```
nit: this could go to `:returns:`

etingof commented on apacheconfig/wloader.py:
```diff
@@ -30,98 +28,149 @@ class Node():
+        :class:`apacheconfig.ItemNode`: ``comment``, ``statement``,
+        ``include``, ``includeoptional``
+
+        :returns: The typestring (as defined by the apacheconfig parser) for
```
Would it make us suffering less if we'd use Napoleon for the docstrings? Vanilla Sphinx parameters syntax seems somewhat verbose and demanding...

FWIW, Napoleon offers two docstring styles to choose from Google and NumPy.

WDYT?

sydneyli replied:
Great suggestion! It's much easier and more readable in code, too-- I gave it a shot in the most recent commit. The branch `sydneyli/apacheconfig:merged-writable-loader-item-sphinx` shows the doc build passing with the napoleon extension.

etingof commented on apacheconfig/wloader.py:
```diff
@@ -30,98 +28,149 @@ class Node():
 
     @whitespace.setter
     def whitespace(self, value):
+        """A string representing literal trailing or preceding whitespace
```
I think Sphinx wants us to have the first line of the docstring short and sweet. We are advised to expand on what's in title in the next paragraph.

etingof commented on apacheconfig/wloader.py:
```diff
@@ -30,98 +28,149 @@ class Node():
         """
         return self._raw[0]
 
     def has_value(self):
```
Can this be a property?

etingof commented on apacheconfig/wloader.py:
```diff
@@ -0,0 +1,219 @@
+
+
+class ItemNode(Node):
+    """Contains data for a comment or option-value directive.
```
I have a vague recollection that the first line of a docstring should be [benevolently imperative](https://www.python.org/dev/peps/pep-0257/).

etingof commented on apacheconfig/wloader.py:
```diff
@@ -0,0 +1,219 @@
+
+
+@six.add_metaclass(abc.ABCMeta)
+class Node():
```
Let's inherit from `object`.

sydneyli commented on apacheconfig/wloader.py:
```diff
@@ -0,0 +1,219 @@
+        """A string representing literal trailing or preceding whitespace
+        for this node. Can be overwritten.
+
+        Each ``ItemNode`` or ``BlockNode`` keeps track of the whitespace
```
Made them all `abstract` & defined them in subclasses so we can put the docstrings in the appropriate places :) that should be easier to maintain going forward.

sydneyli commented:
@etingof this is ready for more review, when you have time ^_^ Biggest change is moving the docs to Napoleon per your suggestion-- the docs are definitely a lot cleaner now.

etingof commented:
> @etingof this is ready for more review, when you have time ^_^ Biggest change is moving the docs to Napoleon per your suggestion-- the docs are definitely a lot cleaner now.

Awesome! Thank you for the heads up - I've been sidetracked by the upcoming conference. I will get back to this shortly.

etingof commented on apacheconfig/wloader.py:
```diff
@@ -0,0 +1,204 @@
+    @abc.abstractmethod
+    def dump(self):
+        """Returns the contents of this node as it would appear in a config
+        file.
```
nit: If this line is wrapped, it better should not. I think Sphinx treats the first line in some special way.

etingof commented on apacheconfig/wloader.py:
```diff
@@ -0,0 +1,204 @@
+    @abc.abstractproperty
+    def whitespace(self):
+        """Returns preceding or trailing whitespace for this node as a string.
+       """
```
nit: how flake8 overlooked wrong indentation?

sydneyli replied:
o.O i'm not sure!

etingof commented on apacheconfig/wloader.py:
```diff
@@ -0,0 +1,204 @@
+        self._whitespace = value
+
+    @staticmethod
+    def parse(raw_str, options={}, parser=None):
```
nit: I think it is generally safe to have immutable sentinels e.g. None.

etingof commented on apacheconfig/wloader.py:
```diff
@@ -0,0 +1,204 @@
+        The "value" is anything but the name. Can be overwritten.
+        """
+        if not self.has_value:
+            return None
```
nit: 'None' is the default, can me omitted

etingof commented on apacheconfig/wloader.py:
```diff
@@ -0,0 +1,204 @@
+            self._raw = self._raw + (" ", value,)
+        self._raw = self._raw[0:-1] + (value,)
+
+    def dump(self):
```
nit: would it make sense to have this as a property like `text` or something? Do you expect some parameters to appear in the future in this function?

sydneyli replied:
It might be good to leave it open to parameters in the future. Either way, I'm wary of making this an attribute in the interface since it actually walks through the entire entire tree to construct the dumped string.

etingof commented on apacheconfig/wloader.py:
```diff
@@ -0,0 +1,204 @@
+        start (str): Which parsing token, as defined by the apacheconfig
+            parser, to expect at the root of strings.
+    """
+    options['preservewhitespace'] = True
```
nit: just noting that another way of overriding the keys is this:

    options.update(preservewhitespace=True, ...)

In case you find it any better.

etingof commented on apacheconfig/wloader.py:
```diff
@@ -0,0 +1,204 @@
+            ``raw_str``.
+        """
+        if not parser:
+            parser = _create_apache_parser(options, start='startitem')
```
I am yet to wrap my head around it, however my concern would be that creating and not reusing the entire stack might be expensive.

sydneyli replied:
Yes, the intention was for folks to create a parser and pass it in! I exposed the default apache parser factory fn and made it a required parameter for `Item.parse`.

etingof commented on apacheconfig/wloader.py:
```diff
@@ -0,0 +1,204 @@
+    """Returns a :class:`apacheconfig.ApacheConfigParser` with default options
+    that are expected by Apache's native parser.
+
+    Overrides apacheconfig options ``preservewhitespace``,
```
What about `namedblocks` option? Should we also disable it to comply with original Apache syntax?

sydneyli replied:
Apache syntax supports named blocks, though they call it something different (arguments rather than block name); for instance, the [Directory](https://httpd.apache.org/docs/2.4/mod/core.html#directory) directive.

etingof commented on apacheconfig/wloader.py:
```diff
@@ -0,0 +1,198 @@
+                       [_restore_original(word) for word in self._raw])))
+
+
+def native_apache_parser(options=None):
```
It seems that this method mainly applies some specific options to lexer/parser/config objects. In some sense, the instantiation code is not relevant to this purpose, right?

So what if we introduce some sort of syntax "flavors" i.e. sets of options, perhaps as global constants, that the user can just pick or mix and match those options to specialize parsing suite to the desired syntax?

sydneyli replied:
Yes, I like the idea of having and exporting a set of common option configurations. Something like `NATIVE_APACHE_PARSER_OPTIONS`?

The motivation behind this was primarily to have a simpler interface for creating a native Apache parser-- would you prefer not to increase the surface area of exported functions to maintain? Either way is fine by me.

etingof replied:
> Yes, I like the idea of having and exporting a set of common option configurations. Something like `NATIVE_APACHE_PARSER_OPTIONS`?

Yeah... BTW, with current design the options are not only for the parser, but for lexer and loader as well.

> The motivation behind this was primarily to have a simpler interface for creating a native Apache parser-- would you prefer not to increase the surface area of exported functions to maintain? Either way is fine by me.

I'd try not to introduce something overly specific and not widely reusable whenever possible.

How about this:

    from apacheconfig import flavors
    from apachaconfig import make_loader

    with make_loader(**flavors.NATIVE_APACHE) as loader:
        ...

If we also need just loader, would it make sense to have a similar function? Or may be we can rename & parameterize `make_loader` so that it would become a universal factory function?

Either way, this factory function(s) would be driven by pre-defined and/or user-specified options.

sydneyli replied:
I'm a fan of exposing the option sets (`flavors`, as you put it) for the time being-- I have to think a bit more on what the factory function might look like, maybe once the other interfaces are in.

etingof replied:
I am thinking, that functionally we specialize parser to work on a specific kind (flavor) of grammar. Each flavor can be composed from a bunch of options.

sydneyli replied:
I gave it a shot with a `flavors.py`-- lmk what you think!

sydneyli commented:
> First of all, I am terribly sorry for long delays! But the good news is that I am hopefully back from my conference preparations!

No worries at all! Thanks for working with us on this :)

> I've posted some more nits, these are the side-effect of my trying to understand the new design. Those are mostly nits, in my mind they might be helpful for the human reader to understand the arrangements. WDYT?

Naming is difficult, so I appreciate your input here! I think they make sense. Took your suggestions.

etingof commented on apacheconfig/flavors.py:
```diff
@@ -0,0 +1,14 @@
+    "multilinehashcomments": True,
+    "useapacheinclude": True,
+}
+""" Set of options for apacheconfig to work with native Apache config files"""
```
nit: leading space

sydneyli commented:
Re: the `make_loader` context manager, could create a generic parser context manager, though probably name it something else (since `make_parser` already exists!) Either way, we could also punt this conversation to another issue for what sort of factory interface to provide for the writable loader.

etingof commented:
> Re: the `make_loader` context manager, could create a generic parser context manager, though probably name it something else (since `make_parser` already exists!) Either way, we could also punt this conversation to another issue for what sort of factory interface to provide for the writable loader.

Alright, so what's remaining for this PR to be merged? My only question so far is possible repetitive parser re-creation. Is this a valid concern?

sydneyli commented:
> Alright, so what's remaining for this PR to be merged? My only question so far is possible repetitive parser re-creation. Is this a valid concern?

That fix should be coming in the downstream PR!

sydneyli commented:
(To be specific, since the LeafNode doesn't contain any other config items, it doesn't need a reference to a parser since it doesn't do any parsing)

etingof commented:
Thank you! \o/

etingof commented:
We should probably update `CHANGES` at some point.

etingof commented:
Another unrelated thought: would it make sense to add "default" flavor essentially enumerating all options and their defaults? That won't technically change anything, but would probably be more illustrative and easier to change defaults.

sydneyli commented:
I think that would be good! a lot of the defaults are scattered in the codebase (using `dict.get(key, default)`) so something like that would consolidate them in one place. I'll open an issue.
